base/threadingconstructs.jl

```diff
@@ -38,8 +38,7 @@ function _threadsfor(iter,lbody)
                 end
             end
             # run this thread's iterations
-            for i = f:l
-                local $(esc(lidx)) = Base.unsafe_getindex(r,i)
+            for $(esc(lidx)) = r[f:l]
```

Comment:
Yes, it will. But that could be remedied by using a iterator expression (eachindex) and reading the value in the loop-expression. 

But then its a conflict between some extra code (and maybe an extra thing to mind while programming) and the possibility of generically parallelising anything with a flat representation.